rfcs/0002-validation/text.md

```diff
@@ -0,0 +1,149 @@
+     from the schema.
+   - Converting the uploaded file to different formats and associating those with the dataset.
+   - Creating a cellxgene deployment for the dataset.
+
+This RFC is concerned with phase 3.
+
+### Relevant Components
+
+There are six main components involved in validation and processing:
+
+![High-level Validation Architecture](imgs/validation_architecture.png)
+
+1. Staging bucket. The S3 bucket where uploaded file are placed by the upload process.
+2. Artifact bucket. The S3 bucket where artifacts associated with datasets are stored and served.
+3. cellxgene deployment bucket. The S3 bucket configured to serve datasets via cellxgene.
+4. Portal database. The database where information about datasets is stored. The portal API queries this database when
+   responding to requests.
+5. Validation and processing image. The Docker image containing the dependencies and script to validate and process.
+6. Validation and processing Fargate task. The task definition for executing the image with AWS Fargate.
+
+#### Portal Database Schema
+
+The relevant part of the portal database is three tables: `dataset`, `artifacts`, and `deployment_directories`.
+
+![Database Schema](imgs/db_schema.png)
+
+The `dataset` table tracks the status of datasets and various metadata about them that is used by the portal frontend.
+There are three status fields:
+
+- upload\_status: uploading, uploaded, canceled
+- validation\_status: validating, valid, invalid
+- artifact\_status: creating, complete
+
+The `artifacts` table tracks the different downloadable files associated with a dataset. This stores the format of the
+download and the bucket and object key in the artifact bucket.
+
+The `deployment_directories` table tracks the cellxgene deployments associated with a dataset and the URL where the
+deployment can be accessed.
+
+### Handoff from Upload
+
+At the end of the upload process there must be two elements present in AWS:
+
+1. A row in the `dataset` table with `id` and `collection_id` populated and `upload_status` set to "uploaded".
+2. A dataset file in the staging bucket with a `x-amz-meta-dataset-id` metadata field whose value is the `id` of the new
+   row in the `dataset` table.
+
+A [CloudWatch Event](https://docs.aws.amazon.com/AmazonCloudWatch/latest/events/CloudWatch-Events-tutorial-ECS.html)
+will trigger the execution of the Fargate task when a new dataset file is added to the staging bucket.
+
+### Steps of the Fargate Task
+
+1. Check the `dataset` table to verify that the expected row exists and `upload_status` is "uploaded". To avoid possible
+   race conditions from the upload process, wait and recheck if this isn't true.
+2. Set the `validation_status` in `dataset` to "validating".
+3. Localize the uploaded file. HDF5 doesn't usually work with remote files very well.
+4. Run the `cellxgene schema validate` command to do a shallow validation of the dataset file.
```

Comment:
Question - if a dataset isn't valid, do we want to keep it? The only way to fix an invalid dataset is by uploading a new one currently so I'm wondering if there should be an cleanup if the dataset fails validation.

Reply:
Does this iteration include[ Implement PII checker for .h5ad datasets](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/corpora-data-portal/376) or it that integrated into [Implement validation in the cellxgene CLI](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/corpora-data-portal/632)?

Reply:
That will get incorporated into the cellxgene CLI.

Reply:
I was thinking we could just stick a pretty short TTL on objects in the upload staging bucket. We just need it there long enough to pull it out and debug if we see something weird. @Bento007 does that seem reasonable to you?

Reply:
And yeah the PII check goes into the CLI. Though we'll have to be a little clever about it, since the signal it's going to give it like "this column looks like birthdays" and maybe the user can override that?

Reply:
Is that in scope for the existing CLI issue of should the data portal PII issue be updated and moved to the cellxgene repo as a follow on.

Reply:
We can configure the object lifecycle using S3 and terraform. We can make it long enough, that validation should be finished before the bucket cleans it up.